pilot/pkg/proxy/envoy/v2/datastructures.go

```diff
@@ -0,0 +1,139 @@
+// 4. Every TCP service in the system will have a TCPRoute (or a filter chain match) associated
+//    with its service IP if present.
+// 5. If a TCP service has no service IP, it will have a listener on the service port and a
+//    catch all TCPRoute
+// 6. Every HTTP port of a service will share the same route configuration (list of virtual hosts)
+//    until further delineation is created by route rules.
+// 7. Similarly, every TCP port of a service will share the same TCPRoute.
+// 7. Routes change (add/delete/update) typically when route rules change
+// 9. Routes can also change when ports are added/removed from the service
+
+// Listeners (react to updates in services, gateways, ext services)
+// 1. Every unique port in the system will have an associated listener.
+// 2. A service can generate one or more listeners based on the number of ports it listens on
+// 3. New listeners are added when a service (incl gateway/external service) with unique
+//    (previously unused) set of ports are added.
+// 4. Listeners are updated when one or more services/gateways/extsvs are added/removed to/from
+//    the same listener port.
+// 5. Listeners are removed when there are no services/gateways/extsvcs listening on the listener
+//    port.
+
+// The indices here are doubly linked. A pair of objects are dependent on each other if they have
+// to be updated when one changes.
+
+// BoxedVirtualHost holds the virtual host proto representation and pointers to
+// owning route configurations.
+type BoxedVirtualHost struct {
+	// The proto representation of this virtual host
+	protoVirtualHost *xdsapi.VirtualHost
+	// list of route configurations (RDS responses) that hold this virtual host
+	routeConfigurations []*BoxedRouteConfiguration
+}
+
+// BoxedRouteConfiguration holds the route configuration proto representation and pointers to
+// owning listeners
+type BoxedRouteConfiguration struct {
+	// The proto representation of this route configuration
+	protoRouteConfiguration *xdsapi.RouteConfiguration
+	// list of virtual hosts attached to this route configuration
+	virtualHosts []*BoxedVirtualHost
+	// list of listeners that use this route configuration
+	listeners []*xdsapi.Listener
+}
+
+// BoxedEndpoint holds the endpoint proto representation and pointers to
+// owning clusters
+type BoxedEndpoint struct {
+	// The proto representation of this endpoint
+	protoEndpoint *xdsapi.LbEndpoint
+	// list of clusters that contain this endpoint
+	clusters []*BoxedCluster
+	// list of labels associated with this endpoint
+	labels []*model.Labels
+}
+
+// BoxedCluster holds the cluster proto representation, pointers to endpoints
+type BoxedCluster struct {
+	// The proto representation of this cluster
+	protoCluster *xdsapi.Cluster
+	// list of endpoints associated with this cluster
+	endpoints []*BoxedEndpoint
+	// List of labels associated with this cluster
+	labels []*model.Labels
```

Comment:
Labels in particular may be better as a map. Don't remember what's inside model.Label, but pointers to structs are sometimes more expensive than a bit of duplicated local data.

Reply:
agreed.